native/src/main/java/io/ballerina/stdlib/data/jsondata/json/JsonCreator.java

```diff
@@ -0,0 +1,243 @@
+                break;
+            case TypeTags.JSON_TAG:
+                nextMapValue = ValueCreator.createMapValue(Constants.JSON_MAP_TYPE);
+                sm.fieldHierarchy.push(new HashMap<>());
+                sm.restType.push(PredefinedTypes.TYPE_JSON);
+                break;
+            case TypeTags.ANYDATA_TAG:
+                nextMapValue = ValueCreator.createMapValue(Constants.ANYDATA_MAP_TYPE);
+                sm.fieldHierarchy.push(new HashMap<>());
+                sm.restType.push(PredefinedTypes.TYPE_JSON);
+                break;
+            default:
+                throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE_FOR_FIELD, getCurrentFieldPath(sm));
+        }
+
+        Object currentJson = sm.currentJsonNode;
+        int valueTypeTag = TypeUtils.getType(currentJson).getTag();
+        if (valueTypeTag == TypeTags.MAP_TAG || valueTypeTag == TypeTags.RECORD_TYPE_TAG) {
+            ((BMap<BString, Object>) currentJson).put(StringUtils.fromString(sm.fieldNames.peek()), nextMapValue);
+        }
+        return Optional.of(nextMapValue);
+    }
+
+    static Optional<BArray> initNewArrayValue(JsonParser.StateMachine sm) {
+        sm.parserContexts.push(JsonParser.StateMachine.ParserContext.ARRAY);
+        Type expType = sm.expectedTypes.peek();
+        if (expType == null) {
+            return Optional.empty();
+        }
+
+        Object currentJsonNode = sm.currentJsonNode;
+        BArray nextArrValue = initArrayValue(sm.expectedTypes.peek());
+        if (currentJsonNode == null) {
+            return Optional.ofNullable(nextArrValue);
+        }
+
+        sm.nodesStack.push(sm.currentJsonNode);
+        return Optional.ofNullable(nextArrValue);
+    }
+
+    private static String getCurrentFieldPath(JsonParser.StateMachine sm) {
+        Iterator<String> itr = sm.fieldNames.descendingIterator();
+
+        StringBuilder result = new StringBuilder(itr.hasNext() ? itr.next() : "");
+        while (itr.hasNext()) {
+            result.append(".").append(itr.next());
+        }
+        return result.toString();
+    }
+
+    static Object convertAndUpdateCurrentJsonNode(JsonParser.StateMachine sm, BString value, Type type) {
+        Object currentJson = sm.currentJsonNode;
+        Object convertedValue = convertToExpectedType(value, type);
+        if (convertedValue instanceof BError) {
+            if (sm.currentField != null) {
+                throw DiagnosticLog.error(DiagnosticErrorCode.INCOMPATIBLE_VALUE_FOR_FIELD, value, type,
+                        getCurrentFieldPath(sm));
+            }
+            throw DiagnosticLog.error(DiagnosticErrorCode.INCOMPATIBLE_TYPE, type, value);
+        }
+
+        Type currentJsonNodeType = TypeUtils.getType(currentJson);
+        switch (currentJsonNodeType.getTag()) {
+            case TypeTags.MAP_TAG:
+            case TypeTags.RECORD_TYPE_TAG:
+                ((BMap<BString, Object>) currentJson).put(StringUtils.fromString(sm.fieldNames.pop()),
```

Comment:
Since this is a record, field names can be not in the proper order.

Reply:
According to the offline discussion. This is not an issue since fieldNames are added to the deque once it matched with parsed key of the `json object`